cranelift/wasm/src/state/module_state.rs

```diff
@@ -21,11 +24,46 @@ pub struct ModuleTranslationState {
     pub(crate) wasm_types: WasmTypes,
 }
 
+fn cranelift_to_wasmparser_type(ty: Type) -> WasmResult<wasmparser::Type> {
+    Ok(match ty {
+        types::I32 => wasmparser::Type::I32,
+        types::I64 => wasmparser::Type::I64,
+        types::F32 => wasmparser::Type::F32,
+        types::F64 => wasmparser::Type::F64,
+        _ => {
+            return Err(WasmError::Unsupported(format!(
+                "Cannot convert Cranelift type to Wasm signature: {:?}",
+                ty
+            )));
+        }
+    })
+}
+
 impl ModuleTranslationState {
     /// Creates a new empty ModuleTranslationState.
     pub fn new() -> Self {
         Self {
             wasm_types: PrimaryMap::new(),
         }
     }
+
+    /// Create a new ModuleTranslationState with the given function signatures,
+    /// provided in terms of Cranelift types. The provided slice of signatures
+    /// is indexed by signature number, and contains pairs of (args, results)
+    /// slices.
+    pub fn from_func_sigs(sigs: &[(Box<[Type]>, Box<[Type]>)]) -> WasmResult<Self> {
```

Comment:
Could the type of the two parameters be `&[Type]`, instead? it would be slightly simpler for users, and wouldn't require the extra call to `into_boxed_slice` on the Spidermonkey side.

Reply:
Done!